torchao/core/config.py

```diff
@@ -26,4 +33,226 @@ def _transform(
+
+        if isinstance(o, dict):
+            return {k: self.encode_value(v) for k, v in o.items()}
+
+        # Default case - let the parent class handle it
+        return super().default(o)
+
+    def encode_value(self, value):
+        """Helper method to recursively encode a value"""
+        # Try to use default for custom type
+        try:
+            # This will handle all our special cases and raise TypeError
+            # if it can't handle the type
+            result = self.default(value)
+            return result
+        except TypeError:
+            pass
+
+        # Default case - return as is
+        # (This will be processed by standard JSON encoder later)
+        return value
+
+
+def config_to_dict(config: AOBaseConfig) -> Dict[str, Any]:
+    """
+    Convert an AOBaseConfig instance to a dictionary suitable for serialization.
+
+    Args:
+        config: An instance of AOBaseConfig subclass
+
+    Returns:
+        Dict representation of the config
+    """
+    if not isinstance(config, AOBaseConfig):
+        raise TypeError(f"Expected AOBaseConfig instance, got {type(config)}")
+
+    # Use the existing JSON encoder but return the dict directly
+    return json.loads(json.dumps(config, cls=ConfigJSONEncoder))
+
+
+ALLOWED_AO_MODULES = {
+    "torchao.quantization",
+}
+
+
+def config_from_dict(data: Dict[str, Any]) -> AOBaseConfig:
+    """
+    Create an AOBaseConfig subclass instance from a dictionary.
+
+    Args:
+        data: Dictionary containing serialized config data
+
+    Returns:
+        An instance of the appropriate AOBaseConfig subclass
+
+    Raises:
+        VersionMismatchError: If the stored version doesn't match the class version
+        ValueError: If deserialization fails for other reasons
+    """
+    if not isinstance(data, dict):
+        raise TypeError(f"Expected dictionary, got {type(data)}")
+
+    if "_type" not in data or "_data" not in data:
+        raise ValueError("Input dictionary missing required '_type' or '_data' fields")
+
+    type_path = data["_type"]
+    stored_version = data.get("_version", 1)
+    obj_data = data["_data"]
+
+    # Handle torch.dtype
+    if type_path == "torch.dtype":
+        import torch
+
+        return getattr(torch, obj_data)
+    # Try to find the class in any of the allowed modules
+    cls = None
+    for module_path in ALLOWED_AO_MODULES:
+        try:
+            module = importlib.import_module(module_path)
+            cls = getattr(module, type_path)
+            break  # Found the class, exit the loop
+        except (ImportError, AttributeError):
+            continue  # Try the next module
+
+    # If we couldn't find the class in any allowed module, raise an error
+    if cls is None:
+        allowed_modules_str = ", ".join(ALLOWED_AO_MODULES)
+        raise ValueError(
+            f"Failed to find class {type_path} in any of the allowed modules: {allowed_modules_str}"
```

Comment:
What's the motivation for this? Why not just make all configs, including custom user configs, serializable?